contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java

```diff
@@ -0,0 +1,928 @@
+ * are evenly distributed among splits (workers). Dataflow checkpointing is fully supported and
+ * each split can resume from previous checkpoint. See
+ * {@link UnboundedKafkaSource#generateInitialSplits(int, PipelineOptions)} for more details on
+ * splits and checkpoint support.
+ *
+ * <pre>{@code
+ * Usage:
+ *   pipeline
+ *     .apply("read_topic_a_and_b",
+ *        KafkaIO.read()
+ *         .withBootstrapServers("broker_1:9092,broker_2:9092")
+ *         .withTopics(ImmutableList.of("topic_a", "topic_b")) // or withTopicPartitions(List<>)
+ *         .withValueCoder(StringUtf8Coder.of())
+ *         .withTimestampFn(timestampFn) // optional
+ *         .withWatermarkFn(watermarkFn)) // optional
+ *     .apply(Values.<String>create()) // discard keys
+ *     ...
+ *}</pre>
+ */
+public class KafkaIO {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaIO.class);
+
+  private static class NowTimestampFn<T> implements SerializableFunction<T, Instant> {
+    @Override
+    public Instant apply(T input) {
+      return Instant.now();
+    }
+  }
+
+  private static class IdentityFn<T> implements SerializableFunction<T, T> {
+    @Override
+    public T apply(T input) {
+      return input;
+    }
+  }
+
+  /**
+   * Creates and uninitialized {@link Read} {@link PTransform}. Before use, basic Kafka
+   * configuration should set with {@link Read#withBootstrapServers(String)} and
+   * {@link Read#withTopics(List)}. Other optional settings include key and value coders,
+   * custom timestamp and watermark functions. Additionally, {@link Read#withMetadata()} provides
+   * access to Kafka metadata for each record (topic name, partition, offset).
+   */
+  public static Read<byte[], byte[]> read() {
+    return new Read<byte[], byte[]>(
+        new ArrayList<String>(),
+        new ArrayList<TopicPartition>(),
+        ByteArrayCoder.of(),
+        ByteArrayCoder.of(),
+        null,
+        null,
+        Read.kafka9ConsumerFactory,
+        Read.defaultConsumerProperties,
+        Long.MAX_VALUE,
+        null);
+  }
+
+  /**
+   * A {@link PTransform} to read from Kafka topics. See {@link KafkaIO#read()} for more
+   * information on configuration.
+   */
+  public static class Read<K, V> extends PTransform<PInput, PCollection<KV<K, V>>> {
+
+    private final List<String> topics;
+    private final List<TopicPartition> topicPartitions; // mutually exclusive with topics
+    private final Coder<K> keyCoder;
+    private final Coder<V> valueCoder;
+    @Nullable private final SerializableFunction<KV<K, V>, Instant> timestampFn;
+    @Nullable private final SerializableFunction<KV<K, V>, Instant> watermarkFn;
+    private final
+      SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>> consumerFactoryFn;
+    private final Map<String, Object> consumerConfig;
+    private final long maxNumRecords; // bounded read, mainly for testing
+    private final Duration maxReadTime; // bounded read, mainly for testing
```

Comment:
drop the "mainly for testing" part -- it may be true as far as we know, but we probably don't want to put value judgements on the user's pipeline decisions.

(Btw, I acknowledge that I think I've written comments that violate what I just wrote in the past. With time comes wisdom.... or so I hope)